packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -51,16 +75,97 @@ const RENDER_APIS: { [primitive in RenderPrimitive]: RenderPrimitiveDefinition }
+        let result = true;
+        const {
+            name: nodeName,
+            namespace = '',
+            attributes,
+            directives,
+            properties,
+            listeners,
+        } = node;
+
+        // Note: the trick is that the static nodes elm can not be removed by themself, they will always be removed
+        //       as parent removal.
+
+        result &&= !isIf(parent); // the parent node is an if, therefore this it is the same as an inline if
+        result &&= !isSlot(node); // slot element can't be static.
+        result &&= !isComponent(node); // components can't be static.
+        // it is an element.
+        result &&= !attributes.some(({ name, value }) => {
+            return (
+                !isLiteral(value) ||
+                name === 'slot' ||
+                // check for ScopedId
+                name === 'id' ||
+                isIdReferencingAttribute(name) ||
+                // Check for ScopedFragId
+                (isSvgUseHref(nodeName, name, namespace) &&
+                    isFragmentOnlyUrl(value.value as string)) ||
+                (isAllowedFragOnlyUrlsXHTML(nodeName, name, namespace) &&
+                    isFragmentOnlyUrl(value.value as string))
+            );
+        }); // all attrs are static
+        result &&= directives.length === 0; // do not have any directive
+        result &&= !properties.some((prop) => !isLiteral(prop.value)); // all properties are static
+        result &&= listeners.length === 0; // do not have any event listener
+
+        // Notes:
+        //   An element inside a foreach can be static because it will have a parent that it is dynamic (the keyed element)
+        //   Slotted content can be static if the parent is dynamic. @todo: think harder in the lightdom cases.
+
+        return result;
+    }
+
+    function collectStaticNodes(node: ChildNode, parent: BaseParentNode) {
+        let childrenAreStatic = true;
+        let nodeIsStatic = true;
+
+        if (isText(node)) {
+            nodeIsStatic = isLiteral(node.value);
+        } else if (isComment(node)) {
+            nodeIsStatic = true;
+        } else {
+            // it is ForBlock | If | BaseElement
+            node.children.forEach((childNode) => {
+                collectStaticNodes(childNode, node);
+
+                childrenAreStatic = childrenAreStatic && staticNodes.has(childNode);
+            });
+
+            nodeIsStatic = isBaseElement(node) && isStaticNode(node, parent);
+        }
+
+        if (nodeIsStatic && childrenAreStatic) {
+            // let's unmark the children as static. So in the codegen we generate the outermost static node.
+            if (isBaseElement(node)) {
+                node.children.forEach((childNode) => staticNodes.delete(childNode));
+            }
+            staticNodes.add(node);
+        }
+    }
+
+    root.children.forEach((childNode) => collectStaticNodes(childNode, root));
+
+    return staticNodes;
+}
+
```

Comment:
Originally I implemented hoisting as a separate step in a separate file (`optimize.ts`) that runs after generation, but I think your approach of doing it directly during the initial generation is superior. It would be great to move `optimize.ts` into this phase as well.

Reply:
I need to think about it, though ill prefer to do it in a different PR.

from the top of my mind, I think there's a semantic difference between the 2 of them: `stc` is an optimization of the generated output, while `hoisted` is changing the generated output based on the input.